syntax = "proto3";

import "proto/mahjong_common.proto";
import "proto/mahjong_rule.proto";

package ycraft.mahjong;

message ScoreCalculatorResult {
  int32 fu = 1;
  int32 fan = 2;
  int32 yakuman = 3;
  repeated Yaku yaku = 4;
  int32 dora = 5;
  int32 uradora = 6;
}

message YakuApplierResult {
  repeated Yaku yaku = 1;
}

message YakuConditionValidatorResult {
  enum Type {
    OK = 0;

    NG_ALLOWED_TILE_CONDITION = -1;
    NG_DISALLOWED_TILE_CONDITION = -2;
    NG_REQUIRED_TILE_CONDITION = -3;
    NG_REQUIRED_ELEMENT_CONDITION = -4;
    NG_REQUIRED_MACHI_TYPE = -5;
    NG_REQUIRED_AGARI_CONDITION = -6;
    NG_REQUIRED_RICHI_TYPE = -7;
    NG_REQUIRED_FIELD_WIND = -8;
    NG_REQUIRED_PLAYER_WIND = -9;

    ERROR_INTERNAL_ERROR = -10;
  }

  Type type = 1;
}

message HandParserResult {
  repeated ParsedHand parsed_hand = 1;
}

message ParsedHand {
  repeated Element element = 1;
  MachiType machi_type = 2;
  Agari agari = 3;
}
